core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -270,27 +323,100 @@ class LogSegment(val log: FileMessageSet,
     catch {
       case e: IOException => throw kafkaStorageException("index", e)
     }
+    try timeIndex.renameTo(new File(CoreUtils.replaceSuffix(timeIndex.file.getPath, oldSuffix, newSuffix)))
+    catch {
+      case e: IOException => throw kafkaStorageException("timeindex", e)
+    }
+  }
+
+  /**
+   * Append the largest time index entry to the time index when this log segment become inactive segment.
+   * This entry will be used to decide when to delete the segment.
+   */
+  def onBecomeInactiveSegment() {
+    timeIndex.maybeAppend(maxTimestampSoFar, offsetOfMaxTimestamp, skipFullCheck = true)
+  }
+
+  /**
+   * The time this segment has waited to be rolled. If the first message in the segment does not have a timestamp,
+   * the time is based on the create time of the segment. Otherwise the time is based on the timestamp of that message.
+   */
+  def timeWaitedForRoll(now: Long) : Long= {
+    // Load the timestamp of the first messsage into memory
+    if (!rollingBaseTimestamp.isDefined) {
+      val iter = log.iterator
+      if (iter.hasNext)
+        rollingBaseTimestamp = Some(iter.next.message.timestamp)
+      else
+        // If the log is empty, we return 0 as time waited.
+        return 0L
```

Comment:
I am thinking that if there is no message in the segment, the log segment would not have been waiting for rolling, so returning 0 seems reasonable. The return value here does not matter much because we check the segment size in Log.maybeRoll(). But it is probably better to return (now - created) since we have a trace level logging in Log.maybeRoll().